modules/table/src/main/java/org/apache/ignite/internal/table/AbstractTableView.java

```diff
@@ -136,24 +154,80 @@ private <T> CompletableFuture<T> withSchemaSync(@Nullable Transaction tx, @Nulla
         return convertToPublicFuture(future);
     }
 
-    private static boolean isOrCausedBy(Class<? extends Exception> exceptionClass, @Nullable Throwable ex) {
-        return ex != null && (exceptionClass.isInstance(ex) || isOrCausedBy(exceptionClass, ex.getCause()));
+    /**
+     * Get index mapping from result set to schema.
+     *
+     * @param columns Columns to map.
+     * @param metadata Metadata for query results.
+     * @return Index mapping.
+     */
+    protected static List<Integer> indexMapping(Column[] columns, ResultSetMetadata metadata) {
+        return Arrays.stream(columns)
+                .map(Column::name)
+                .map(IgniteNameUtils::quoteIfNeeded)
+                .map((columnName) -> {
+                    int rowIdx = metadata.indexOf(columnName);
+
+                    if (rowIdx == -1) {
+                        throw new IgniteException(Common.INTERNAL_ERR, "Missing required column in query results: " + columnName);
+                    }
+
+                    return rowIdx;
+                })
+                .collect(toList());
     }
 
     /**
-     * Construct SQL query and arguments for prepare statement.
+     * Create conversion function for objects contained by result set to criteria query objects.
      *
-     * @param tableName Table name.
-     * @param columnNames Column names.
-     * @param criteria The predicate to filter entries or {@code null} to return all entries from the underlying table.
-     * @return SQL query and it's arguments.
+     * @param meta Result set columns' metadata.
+     * @param schema Schema.
+     * @return Conversion function (if {@code null} conversions isn't required).
      */
-    static SqlSerializer createSqlSerializer(String tableName, Collection<String> columnNames, @Nullable Criteria criteria) {
-        return new SqlSerializer.Builder()
-                .tableName(tableName)
-                .columns(columnNames)
-                .where(criteria)
-                .build();
+    protected @Nullable Function<SqlRow, R> queryMapper(ResultSetMetadata meta, SchemaDescriptor schema) {
+        return null;
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public Cursor<R> query(@Nullable Transaction tx, @Nullable Criteria criteria, CriteriaQueryOptions opts) {
+        return new CursorAdapter<>(sync(queryAsync(tx, criteria, opts)));
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public CompletableFuture<AsyncCursor<R>> queryAsync(
+            @Nullable Transaction tx,
+            @Nullable Criteria criteria,
+            CriteriaQueryOptions opts
+    ) {
+        CriteriaQueryOptions opts0 = opts == null ? CriteriaQueryOptions.DEFAULT : opts;
```

Comment:
if `opts` can be null it should be marked as `@Nullable`

Reply:
Oops, this was missed after merge